mdnsd/packet.c

```diff
@@ -1115,6 +1055,70 @@ pkt_parse_rr(u_int8_t **pbuf, u_int16_t *len, struct rr *rr)
 	case T_AAAA:
 	case T_NSEC:
 		break;
+	case T_OPT:
+		u_int16_t i, code, plen, erc, pl;
+		size_t j;
+
+		/* 
+		 * We need 8 bytes for the OPT RR frame.
+		 * See RFC 2671, 4.3.
+		 */
+		if (*len < 8) {
+			log_warnx("Unexpected packet len "
+				"(T_OPT header malformed)");
+			return (-1);
+		}
+
+		/* sender's UDP payload size */
+		GETSHORT(pl, *pbuf);
+		*len -= INT16SZ;
+		/* extended RCODE and flags */
+		GETLONG(erc, *pbuf);
+		*len -= INT32SZ;
+		/* describes RDATA */
+		GETSHORT(us, *pbuf);
+		*len -= INT16SZ;
+		if (*len < us) {
+			log_warnx("Unexpected packet len "
+				"(T_OPT data malformed)");
+			return (-1);
+		}
+
+		for (j = 0, i = 0; i < us; j++) {
+			GETSHORT(code, *pbuf);
+			*len -= INT16SZ;
+			i += INT16SZ;
+			GETSHORT(plen, *pbuf);
+			*len -= INT16SZ; 
+			i += INT16SZ;
+			if (us - i < plen) {
+				log_warnx("Unexpected packet len "
+					"(T_OPT section %zu malformed)",
+					j);
+				return (-1);
+			}
+
+			if (4 == code) {
+				/* draft-cheshire-edns0-owner-option-00.txt */
+				log_debug("pkt_parse_rr: %s: "
+					"EDNSD0 \'Owner\'",
+					rr_type_name(rr->rrs.type));
+			} else {
+				log_warnx("pkt_parse_rr: %s: %llu "
+					"bytes in RDATA section %zu, "
+					"unknown code %llu",
+					rr_type_name(rr->rrs.type),
+					(long long unsigned)plen, j,
+					(long long unsigned)code);
+			}
+			*pbuf += plen;
+			*len -= plen;
+			i += plen;
+		}
+
+		rr->flags = 0;
+		rr->ttl = 0;
+		break;
```

Comment:
I just can't figure out where the `Unprocessed rr in Additional Section` comes in to play after it processes the EDNS0 packets.